src/server/main_service.cc

```diff
@@ -1565,52 +1564,70 @@ void Service::EvalInternal(const EvalArgs& eval_args, Interpreter* interpreter,
 
   DCHECK(!cntx->conn_state.script_info);  // we should not call eval from the script.
 
+  optional<ShardId> sid;
+
   // TODO: to determine whether the script is RO by scanning all "redis.p?call" calls
   // and checking whether all invocations consist of RO commands.
   // we can do it once during script insertion into script mgr.
   auto& sinfo = cntx->conn_state.script_info;
   sinfo = make_unique<ConnectionState::ScriptInfo>();
   for (size_t i = 0; i < eval_args.keys.size(); ++i) {
-    sinfo->keys.insert(KeyLockArgs::GetLockKey(ArgS(eval_args.keys, i)));
+    string_view key = KeyLockArgs::GetLockKey(ArgS(eval_args.keys, i));
+    sinfo->keys.insert(key);
+
+    ShardId cur_sid = Shard(key, shard_count());
+    if (i == 0) {
+      sid = cur_sid;
+    }
+    if (sid.has_value() && *sid != cur_sid) {
+      sid = nullopt;
+    }
   }
+
```

Comment:
The only thing I'm worried about is that we compute hashes twice now. I remember when I was benchmarking tx code a few months ago, computing Shard() actually took up to 5% of execution time.

What we could do in the _future_, is for example avoid Scheudling in `StartMultiLockedAhead`. And then also re-use this function for MULTI/EXEC. This way we don't need to compute hashes and re-initialize the transaction

Reply:
That's a good idea!